tests/php/test-class-amp-theme-support.php

```diff
@@ -1289,6 +1289,213 @@ function() {
 		$this->assertNotContains( 'no-customize-support', $body_classes );
 	}
 
+	/**
+	 * Assert that dev mode attribute *is* on the queried element.
+	 *
+	 * @param DOMXPath $xpath XPath.
+	 * @param string   $query Query.
+	 */
+	public function assert_dev_mode_is_on_queried_element( DOMXPath $xpath, $query ) {
+		$element = $xpath->query( $query )->item( 0 );
+		$this->assertInstanceOf( 'DOMElement', $element, 'Expected element for query: ' . $query );
+		$this->assertTrue( $element->hasAttribute( AMP_Rule_Spec::DEV_MODE_ATTRIBUTE ), 'Expected dev mode to be enabled on element for query: ' . $query );
+	}
+
+	/**
+	 * Assert that dev mode attribute is *not* on the queried element.
+	 *
+	 * @param DOMXPath $xpath XPath.
+	 * @param string   $query Query.
+	 */
+	public function assert_dev_mode_is_not_on_queried_element( DOMXPath $xpath, $query ) {
+		$element = $xpath->query( $query )->item( 0 );
+		$this->assertInstanceOf( 'DOMElement', $element, 'Expected element for query: ' . $query );
+		$this->assertFalse( $element->hasAttribute( AMP_Rule_Spec::DEV_MODE_ATTRIBUTE ), 'Expected dev mode to not be enabled on element for query: ' . $query );
+	}
+
+	/**
+	 * Get data to test AMP_Theme_Support::filter_admin_bar_style_loader_tag().
+	 *
+	 * @return array
+	 */
+	public function get_data_to_test_filtering_admin_bar_style_loader_tag_data() {
+		return [
+			'admin_bar_exclusively_dependent'             => [
+				static function () {
+					wp_enqueue_style( 'example-admin-bar', 'https://example.com/example-admin-bar.css', [ 'admin-bar' ], '0.1' );
+				},
+				function ( DOMXPath $xpath ) {
+					$this->assert_dev_mode_is_on_queried_element( $xpath, '//link[ @id = "example-admin-bar-css" ]' );
+					$this->assert_dev_mode_is_on_queried_element( $xpath, '//link[ @id = "dashicons-css" ]' );
+				},
+			],
+
+			'dashicons_enqueued_independent_of_admin_bar' => [
+				static function () {
+					wp_enqueue_style( 'dashicons' );
+				},
+				function ( DOMXPath $xpath ) {
+					$this->assert_dev_mode_is_not_on_queried_element( $xpath, '//link[ @id = "dashicons-css" ]' );
+				},
+			],
+
+			'dashicons_enqueued_independent_of_also_enqueued_admin_bar' => [
+				static function () {
+					wp_enqueue_style( 'admin-bar' );
+					wp_enqueue_style( 'dashicons' );
+				},
+				function ( DOMXPath $xpath ) {
+					$this->assert_dev_mode_is_not_on_queried_element( $xpath, '//link[ @id = "dashicons-css" ]' );
+				},
+			],
+
+			'dashicons_not_dev_mode_because_non_admin_bar_dependency' => [
+				static function () {
+					wp_enqueue_style( 'admin-bar' );
+					wp_enqueue_style( 'special-icons', 'https://example.com/special-icons.css', [ 'dashicons' ], '0.1' );
+				},
+				function ( DOMXPath $xpath ) {
+					$this->assert_dev_mode_is_not_on_queried_element( $xpath, '//link[ @id = "dashicons-css" ]' );
+				},
+			],
+
+			'dashicons_in_dev_mode_because_all_dependents_depend_on_admin_bar' => [
```

Comment:
This is the case that I think is debatable. I don't think that having both `'dashicons'` and `'admin-bar'` as dependencies warrants the conclusion that `dashicons` is not used outside the admin bar.

Reply:
I'm not sure. The `special-icons` style here depends on `admin-bar` so if you enqueue it then you'd also get the `admin-bar`. In other words, it's true that perhaps some of the icons are being used outside the admin bar area, but perhaps this is because of some inline editing features. This is good because those special icons are needed by that authenticated user in dev mode. So if the user is logged-out, and the admin bar is not being included, then these should not be getting enqueued.